docs/architecture/adr-023-gas-used-and-gas-price-estimation.md

```diff
@@ -0,0 +1,113 @@
+# ADR 023: Gas used and gas price estimation
+
+## Changelog
+
+- 2025/01/17: Initial draft (@rach-id)
+
+## Status
+
+Proposed
+
+## Context
+
+As per [CIP-18: Standardised Gas and Pricing Estimation Interface](https://github.com/celestiaorg/CIPs/blob/main/cips/cip-18.md), celestia-app will provide a base implementation for the gas and fee estimation so it can be used when building transactions.
+
+## Decision
+
+We will proceed with the implementation in celestia-app.
+
+The fee estimation mechanism is described below.
+
+## Detailed Design
+
+The API will be provided through gRPC following the interface provided in [CIP-18](https://github.com/celestiaorg/CIPs/blob/main/cips/cip-18.md).
+
+### Gas used estimation
+
+The gas used estimation will be calculated using the baseapp simulation mode:
+
+```go
+	gasInfo, _, err := app.BaseApp.Simulate(tx)
+	if err != nil {
+		return nil, err
+	}
+```
+
+_Note:_ The returned result will exactly match the output of the state machine, without any gas multiplier applied to adjust the values.
+Clients using this endpoint are responsible for accounting for any necessary multipliers.
+
+### Gas price estimation
+
+There are multiple ways to have a gas price estimation for a priority level. For a first implementation, we will use the following:
+
+- High Priority: The gas price is the price at the start of the top 10% of transactions’ gas prices from the last 5 blocks.
+- Medium Priority: The gas price is the median of all gas prices from the last 5 blocks.
+- Low Priority: The gas price is the value at the end of the lowest 10% of gas prices from the last 5 blocks.
+- None Priority: This is equivalent to the Medium priority, using the median of all gas prices from the last 5 blocks.
```

Comment:
Just something to be wary of: this doesn't consider the fullness of the blocks. For example if there is only one transaction in the last 5 blocks, The new pricing will be in direct relation of what the price of that transaction was. Yet if the blocks continue to be empty, the estimation should be the minimum

Reply:
yes, in this case, the returned gas price will be that of that transaction. There are many edge cases we can consider in this and multiple ways to enhance this estimation. 

The most important IMO is to return a gas price estimate > than the minimum, only in the case of the last 5 blocks are all > 50% full or some other number.

But I guess we just want to provide an API for now, even if the numbers are skewed, to lay ground for the community to build more accurate estimation algorithms.

If not, we can rework the estimation mechanism.